### PR TITLE
[transform.basicprofiles] Debounce (time) profile

### PR DESCRIPTION
**Pull request by J-N-K**
This adds a second strategy for debouncing values. If applied to a link, the state/commands are only forwarded if no furter update is received during the configured timespan.

### CODE REVIEW

File: bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/factory/BasicProfilesFactory.java [hunk 1]
```diff
@@ -92,16 +98,18 @@
...
-            PROFILE_TYPE_THRESHOLD);
+            PROFILE_TYPE_GENERIC_TOGGLE_SWITCH, PROFILE_TYPE_DEBOUNCE_COUNTING, PROFILE_TYPE_DEBOUNCE_TIME,
+            PROFILE_TYPE_INVERT, PROFILE_TYPE_ROUND, PROFILE_TYPE_THRESHOLD);
 
     private final Map<LocalizedKey, ProfileType> localizedProfileTypeCache = new ConcurrentHashMap<>();
 
     private final ProfileTypeI18nLocalizationService profileTypeI18nLocalizationService;
     private final Bundle bundle;
 
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool("basicprofile");
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/factory/BasicProfilesFactory.java:111]: Do you create an dedicated thread pool on purpose? The `ProfileContext` already provides its own thread pool via `getExecutorService()` method.

```suggestion
    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool("basicprofiles");
```

File: bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml [hunk 2]
```diff
@@ -11,4 +11,17 @@
 			<default>1</default>
 		</parameter>
 	</config-description>
+
+	<config-description uri="profile:basic-profiles:debounce-time">
+		<parameter name="toItemDelay" type="integer" min="0" step="1">
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:16]: You may want to add the unit attribute here:

```suggestion
		<parameter name="toItemDelay" type="integer" min="0" step="1" unit="ms">
```

File: bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml [hunk 3]
```diff
@@ -11,4 +11,17 @@
...
+			<label>To Item Delay</label>
+			<description>Number of changes before updating Item State.</description>
+			<default>0</default>
+		</parameter>
+		<parameter name="toHandlerDelay" type="integer" min="0" step="1">
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:21]:
```suggestion
		<parameter name="toHandlerDelay" type="integer" min="0" step="1" unit="ms">
```

File: bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/i18n/basicprofiles.properties [hunk 4]
```diff
@@ -9,8 +9,14 @@ profile.config.basic-profiles.toggle-switch.events.label = Events
 profile.config.basic-profiles.toggle-switch.events.description = Comma separated list of events to which the profile should listen.
 
 profile-type.basic-profiles.debounce-counting.label = Debounce (Counting)
-profile.config.basic-profiles.debounce-counting.scale.label = Number Of Changes
-profile.config.basic-profiles.debounce-counting.scale.description = Number of changes before updating Item State.
+profile.config.basic-profiles.debounce-counting.numberOfChanges.label = Number Of Changes
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/i18n/basicprofiles.properties:12]: :+1: Thanks for fixing this.

File: bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/config/DebounceTimeStateProfileConfig.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.smarthomej.transform.basicprofiles.internal.config;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.smarthomej.transform.basicprofiles.internal.profiles.DebounceCountingStateProfile;
+
+/**
+ * Configuration for {@link DebounceCountingStateProfile}.
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/config/DebounceTimeStateProfileConfig.java:19]:
```suggestion
 * Configuration for {@link DebounceTimeStateProfile}.
```

**Review by cweitkamp — CHANGES_REQUESTED** [review]
Looks good so far. I added some remarks.

We now could add a `Throttle` Profile too, which accepts a state/command first and blocks following state changes/commands for a user defined amount of time.

File: bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/factory/BasicProfilesFactory.java (hunk 1 above)
Comment (J-N-K) [reply bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/factory/BasicProfilesFactory.java:111]: Didn't see that. Removed own threadpool

**Comment by J-N-K** [comment]
I think I addresses all comments. The "throttle" could be easily integrated by adding a "send first/send last" parameter to the profile, I'll check that later.

File: bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+    public DebounceTimeStateProfile(ProfileCallback callback, ProfileContext context) {
+        this.callback = callback;
+        this.scheduler = context.getExecutorService();
+        this.config = context.getConfiguration().as(DebounceTimeStateProfileConfig.class);
+        logger.debug("Configuring profile with parameters: {}", config);
+
+        if (config.toHandlerDelay < 0) {
+            throw new IllegalArgumentException(String.format(
+                    "debounceTimeToHandler has to be a non-negative integer but was '%d'.", config.toHandlerDelay));
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java:60]:
```suggestion
                    "'toHandlerDelay' has to be a non-negative integer but was '%d'.", config.toHandlerDelay));
```

File: bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        if (config.toItemDelay < 0) {
+            throw new IllegalArgumentException(String
+                    .format("debounceTimeToItem has to be a non-negative integer but was '%d'.", config.toItemDelay));
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java:65]:
```suggestion
                    .format("'toItemDelay' has to be a non-negative integer but was '%d'.", config.toItemDelay));
```

File: bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml [hunk 8]
```diff
@@ -11,4 +11,26 @@
 			<default>1</default>
 		</parameter>
 	</config-description>
+
+	<config-description uri="profile:basic-profiles:debounce-time">
+		<parameter name="toItemDelay" type="integer" min="0" step="1" unit="ms">
+			<label>To Item Delay</label>
+			<description>Number of changes before updating Item State.</description>
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:18]: Can you change it to a meaningful explanation for the parameter?

File: bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml [hunk 9]
```diff
@@ -11,4 +11,26 @@
...
+			<default>0</default>
+		</parameter>
+		<parameter name="toHandlerDelay" type="integer" min="0" step="1" unit="ms">
+			<label>To Handler Delay</label>
+			<description>Number of changes before updating Item State.</description>
```
Comment (cweitkamp) [inline bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:23]: See above.

**Review by cweitkamp — COMMENTED** [review]
Looks promising. Thanks. Some final remarks.

**Review by cweitkamp — APPROVED** [review]
LGTM.

**Pull request merged**